### PR TITLE
[nginx-ingress-controller] Custom errors should be optional

### LINKED ISSUE #897: Nginx Ingress Controller - Option to disable interception of errors
It seems to me right now there is no way to configure the nginx ingress controller to **not** intercept errors and serve them from the default/error backend.

In our case this is problematic. Our backends serve meaningful error pages for 500 errors which contain exception ids which help us debugging problems when customers create support requests.
We also have a JSON API backend where 404 error contain additional information that help api clients to fail more gracefully.

How about adding an option to (potentially partially) disable the *error interception* ?
I can submit an initial PR of this would be considered useful.

### PR DESCRIPTION
**Pull request by aledbf**
Without this change the errors in the upstream server are redirected to the default backend.

fixes #897

### CODE REVIEW

**Comment by databus23** [comment]
Any progress on this? I'm eagerly waiting for this. :)
/cc @bprashanth

File: ingress/controllers/nginx/nginx/main.go [hunk 1]
```diff
@@ -124,6 +124,20 @@ type nginxConfiguration struct {
 	// accessed using HTTPS.
 	HSTSMaxAge string `structs:"hsts-max-age,omitempty"`
 
+	// enables or disable if HTTP codes should be passed to a client or be redirected to nginx for
```
Comment (bprashanth) [inline ingress/controllers/nginx/nginx/main.go]: Hmm, can you just pass a list and parse in the controller instead of each code?

Comment (bprashanth) [inline ingress/controllers/nginx/nginx/main.go]: can you also add some text to the docs about this?

Comment (aledbf) [inline ingress/controllers/nginx/nginx/main.go]: done

File: ingress/controllers/nginx/nginx/template.go [hunk 2]
```diff
@@ -56,15 +56,9 @@ func (ngx *Manager) writeCfg(cfg nginxConfiguration, ingressCfg IngressConfig) (
...
 	conf["cfg"] = fixKeyNames(structs.Map(cfg))
 
-	buffer := new(bytes.Buffer)
-	err := ngx.template.Execute(buffer, conf)
-	if err != nil {
-		glog.Infof("NGINX error: %v", err)
-		return false, err
-	}
-
 	if glog.V(3) {
```
Comment (bprashanth) [inline ingress/controllers/nginx/nginx/template.go]: suggest simply having a global --verbose=true and flipping log level to V(4) everywhere. That's been helpful for me at least, I almost never really want levelled logs. Just everything, or default. If you want to do that (not in this pr) you can copy how https://github.com/kubernetes/contrib/blob/master/ingress/controllers/gce/main.go#L111 is handled

File: ingress/controllers/nginx/nginx/utils.go [hunk 3]
```diff
@@ -91,6 +96,19 @@ func (ngx *Manager) ReadConfig(config *api.ConfigMap) nginxConfiguration {
 		Metadata:         metadata,
 	})
 
+	cErrors := make([]int, 0)
+	if val, ok := config.Data[customHTTPErrors]; ok {
+		delete(config.Data, customHTTPErrors)
+		for _, i := range strings.Split(val, ",") {
+			j, err := strconv.Atoi(i)
+			if err != nil {
+				glog.Warningf("%v is not a valid http code", i)
```
Comment (bprashanth) [inline ingress/controllers/nginx/nginx/utils.go]: all you can say is that this is not an int, right? can you also log the error?

File: ingress/controllers/nginx/nginx/utils.go [hunk 4]
```diff
@@ -115,9 +133,23 @@ func (ngx *Manager) ReadConfig(config *api.ConfigMap) nginxConfiguration {
...
 	}
 
+	cfgDefault.CustomHTTPErrors = ngx.filterErrors(cErrors)
 	return cfgDefault
 }
 
+func (ngx *Manager) filterErrors(errCodes []int) []int {
+	fa := make([]int, 0)
+	for _, errCode := range errCodes {
+		if errCode > 299 && errCode < 600 {
```
Comment (bprashanth) [inline ingress/controllers/nginx/nginx/utils.go]: is this a continguous range?

**Comment by bprashanth** [comment]
LGTM sorry for the delay, all I had were a couple of nits that can easily be fixed in a follow up pr

File: ingress/controllers/nginx/nginx/utils.go (hunk 4 above)
Comment (aledbf) [inline ingress/controllers/nginx/nginx/utils.go]: yes, any code in that range can be customized

File: ingress/controllers/nginx/nginx/utils.go (hunk 3 above)
Comment (aledbf) [inline ingress/controllers/nginx/nginx/utils.go]: ok, duly noted for the next PR

**Pull request merged**